iina/FFmpegController.mm

```diff
@@ -0,0 +1,205 @@
+//
+//  FFmpegController.mm
+//  iina
+//
+//  Created by Saagar Jha on 2/16/19.
+//  Copyright © 2019 lhc. All rights reserved.
+//
+
+#import "FFmpegController.h"
+
+// Perhaps this will work someday? Until then, we need to keep around our dummy
+// interface in the header file.
+//#import "IINA-Swift.h"
+
+#import <algorithm>
+#import <memory>
+
+extern "C" {
+#import <libavcodec/avcodec.h>
+#import <libavformat/avformat.h>
+#import <libavutil/avutil.h>
+#import <libavutil/imgutils.h>
+#import <libswscale/swscale.h>
+}
+
+#define DEFAULT_THUMBNAIL_COUNT 100
+#define DEFAULT_THUMBNAIL_WIDTH 240.0
+
+#define __ACTUALLY_CONCATENATE(a, b) a##b
+#define __CONCATENATE(a, b) __ACTUALLY_CONCATENATE(a, b)
+#define DEFER(cleanup) std::shared_ptr<void> __CONCATENATE(__defer_, __LINE__)(nullptr, cleanup)
+
+@implementation FFmpegController (Bridge)
+
+- (BOOL)synchronouslyGenerateThumbnailsForFileAtPath:(NSString *)path {
+  AVFormatContext *formatContext = NULL;
+  if (avformat_open_input(&formatContext, path.fileSystemRepresentation, nullptr, nullptr) < 0) {
+    return NO;
+  }
```

Comment:
I personally prefer now to write optional parentheses in C&C++. We do have some if statements that simply return `NO`, so I think remove those parentheses might be better in terms of readability.

Reply:
Are you talking about the brackets?

Reply:
Yeah, sorry about that.

Reply:
Actually, now that we have a reliable cleanup mechanism I'm thinking that we should bring back the macros to handle these.